src/OAuth2/Provider/SmashCast.php

```diff
@@ -0,0 +1,215 @@
+    {
+        if (empty($body)) {
+            throw new InvalidAccessToken('Provider response with empty body');
+        }
+
+        $result = json_decode($body, true);
+        if ($result) {
+            return new AccessToken($result);
+        }
+
+        throw new InvalidAccessToken('Provider response with not valid JSON');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function makeAccessTokenRequest($code)
+    {
+        $parameters = [
+            'request_token' => $code,
+            'app_token' => $this->consumer->getKey(),
+            'hash' => base64_encode($this->consumer->getKey() . $this->consumer->getSecret()),
+        ];
+
+        return new \SocialConnect\Common\Http\Request(
+            $this->getRequestTokenUri(),
+            $parameters,
+            $this->requestHttpMethod,
+            [
+                'Content-Type' => 'application/x-www-form-urlencoded'
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAccessTokenByRequestParameters(array $parameters)
+    {
+        $state = $this->session->get('oauth2_state');
+        if (!$state) {
+            throw new UnknownAuthorization();
+        }
+
+        if (!isset($parameters['state'])) {
+            throw new UnknownState();
+        }
+
+        if ($state !== $parameters['state']) {
+            throw new InvalidState();
+        }
+
+        if (isset($parameters['authToken'])) {
+            return new AccessToken(['access_token' => $parameters['authToken']]);
+        }
+
+        return $this->getAccessToken($parameters['request_token']);
+    }
+
+    /**
+     * This method it needed, because I cannot fix auth/login with accessToken
+     * BTW: Yes, I known that it's unneeded round trip to the server
+     *
+     * @param AccessTokenInterface $accessToken
+     * @return mixed
+     * @throws InvalidResponse
+     */
+    protected function getUserNameByToken(AccessTokenInterface $accessToken)
+    {
+        $response = $this->httpClient->request(
+            $this->getBaseUri() . 'userfromtoken/' . $accessToken->getToken()
+        );
+
+        if (!$response->isSuccess()) {
+            throw new InvalidResponse(
+                'API response with error code',
+                $response
+            );
+        }
+
+        $result = $response->json();
+
+        if (!$result) {
+            throw new InvalidResponse(
+                'API response is not a valid JSON object',
+                $response->getBody()
+            );
+        }
+
+        return $result->user_name;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getIdentity(AccessTokenInterface $accessToken)
+    {
+        // @todo Find a problem with this code
+        /*$response = $this->httpClient->request(
+            $this->getBaseUri() . 'auth/login',
+            [
+                'app' => 'desktop', // @any app name, not working, I was using JSON and not working..
```

Comment:
@jens1o Should i use `socialconnect`, OAuth app name? or something else?

![image](https://user-images.githubusercontent.com/572096/27436038-dfac1292-5798-11e7-93ec-5aaefd51e48c.png)